cura/Settings/MachineManager.py

```diff
@@ -1496,3 +783,268 @@ def setSettingForAllExtruders(self, setting_name: str, property_name: str, prope
         for key, extruder in self._global_container_stack.extruders.items():
             container = extruder.userChanges
             container.setProperty(setting_name, property_name, property_value)
+
+    #
+    # New
+    #
+
+    # We not fetch it from _current_root_material_id, but later we can get it from somewhere else
+    @pyqtProperty("QVariantList", notify = rootMaterialChanged)
+    def currentExtruderPositions(self):
+        return sorted(list(self._current_root_material_id.keys()))
+
+    @pyqtProperty("QVariant", notify = rootMaterialChanged)
+    def currentRootMaterialId(self):
+        # initial filling the current_root_material_id
+        self._current_root_material_id = {}
+        for position in self._global_container_stack.extruders:
+            self._current_root_material_id[position] = self._global_container_stack.extruders[position].material.getMetaDataEntry("base_file")
+        return self._current_root_material_id
+
+    @pyqtProperty("QVariant", notify = rootMaterialChanged)
+    def currentRootMaterialName(self):
+        # initial filling the current_root_material_name
+        if self._global_container_stack:
+            self._current_root_material_name = {}
+            for position in self._global_container_stack.extruders:
+                if position not in self._current_root_material_name:
+                    material = self._global_container_stack.extruders[position].material
+                    self._current_root_material_name[position] = material.getName()
+        return self._current_root_material_name
+
+    ##  Return the variant names in the extruder stack(s).
+    ##  For the variant in the global stack, use activeVariantBuildplateName
+    @pyqtProperty("QVariant", notify = activeVariantChanged)
+    def activeVariantNames(self):
+        result = {}
+
+        active_stacks = ExtruderManager.getInstance().getActiveExtruderStacks()
+        if active_stacks is not None:
+            for stack in active_stacks:
+                variant_container = stack.variant
+                position = stack.getMetaDataEntry("position")
+                if variant_container and variant_container != self._empty_variant_container:
+                    result[position] = variant_container.getName()
+
+        return result
+
+    def _setEmptyQuality(self):
```

Comment:
Needs documentation... What does this do?

Reply:
Added doc in 413e4ca55